### PR TITLE
Enable splines for LODE

### PR DESCRIPTION
**Pull request by PicoCentauri**
This is a WIP work and currently works but doesn't support the `central_atom_contribution`

### CODE REVIEW

File: rascaline/src/calculators/lode/radial_integral/spline.rs [hunk 1]
```diff
@@ -63,6 +64,33 @@ impl LodeRadialIntegralSpline {
...
+                    position: spline_point.position,
+                    values: spline_point.values.0.clone(),
+                    derivatives: spline_point.derivatives.0.clone(),
+                }
+            );
+        }
+
+        let spline = HermitCubicSpline::new(spline_parameters, new_spline_points);
+        return Ok(LodeRadialIntegralSpline{
+            spline: spline, center_contribution: Array1::zeros(parameters.max_radial)});
```
Comment (PicoCentauri) [inline rascaline/src/calculators/lode/radial_integral/spline.rs:92]: `center_contribution` is currently set to zero here, we will also have to provide these values with the splines something like:

```python
radial_basis={"TabulatedRadialIntegral": {"points": spline_points, "center_contribution": xxx}},
```

File: rascaline/src/calculators/radial_basis/mod.rs [hunk 2]
```diff
@@ -28,7 +28,12 @@ pub enum RadialBasis {
     /// The easiest way to create a set of spline points is the
     /// `rascaline.generate_splines` Python function.
     TabulatedRadialIntegral {
+        /// spline points
         points: Vec<SplinePoint>,
+        /// contribution of the central atom. This parameter is only required for LODE
+        /// calculations where the length of `center_contribution` must be the same as
+        /// `max_radial`.
+        center_contribution: Option<Vec<f64>>,
```
Comment (PicoCentauri) [inline rascaline/src/calculators/radial_basis/mod.rs:36]: I made the `center_contribution` an optional argument. Let me know if this is okay of if we should create a new enum entry like `TabulatedRadialIntegralCenter`

File: rascaline/src/calculators/lode/radial_integral/spline.rs [hunk 3]
```diff
@@ -63,6 +64,40 @@ impl LodeRadialIntegralSpline {
...
+            new_spline_points.push(
+                HermitSplinePoint{
+                    position: spline_point.position,
+                    values: spline_point.values.0.clone(),
+                    derivatives: spline_point.derivatives.0.clone(),
+                }
+            );
+        }
+
+        assert_eq!(
```
Comment (Luthaf) [inline rascaline/src/calculators/lode/radial_integral/spline.rs:91]: I think this should be a `InvalidParameterError`, `assert_eq` will `panic`, which creates an `internal error (this is a bug)` error message

File: rascaline/src/calculators/lode/radial_integral/spline.rs [hunk 4]
```diff
@@ -63,6 +64,40 @@ impl LodeRadialIntegralSpline {
             center_contribution: radial_integral.compute_center_contribution()
         });
     }
+
+    pub fn from_tabulated(
```
Comment (Luthaf) [inline rascaline/src/calculators/lode/radial_integral/spline.rs:68]: Please add a short doc comment here

File: rascaline/src/calculators/radial_basis/mod.rs (hunk 2 above)
Comment (Luthaf) [reply rascaline/src/calculators/radial_basis/mod.rs:36]: That's fine for me, but we need to document a bit better what "contribution of the central atom" means (i.e. what's the equation that should be computed)

File: rascaline/src/calculators/lode/radial_integral/spline.rs (hunk 4 above)
Comment (PicoCentauri) [reply rascaline/src/calculators/lode/radial_integral/spline.rs:68]: Done!

**Comment by PicoCentauri** [comment]
I added the equation and applied some extra math typestting.

File: rascaline/src/calculators/radial_basis/mod.rs [hunk 5]
```diff
@@ -10,8 +10,8 @@ pub use self::tabulated::SplinePoint;
 pub enum RadialBasis {
     /// Use a radial basis similar to Gaussian-Type Orbitals.
     ///
-    /// The basis is defined as `R_n(r) ∝ r^n e^{- r^2 / (2 σ_n^2)}`, where `σ_n
-    /// = cutoff * \sqrt{n} / n_max`
+    /// The basis is defined as $R_n(r) \propto r^n \exp\left[ -r^2 / (2 σ_n^2) \right]$,
```
Comment (Luthaf) [inline rascaline/src/calculators/radial_basis/mod.rs:13]: This is not rendered as math, so I think the previous code is better here: https://rascaline--204.org.readthedocs.build/en/204/references/calculators/spherical-expansion.html#json-schema-radialbasis

File: rascaline/src/calculators/radial_basis/mod.rs [hunk 6]
```diff
@@ -28,7 +28,22 @@ pub enum RadialBasis {
...
+        ///        \sqrt{4 \pi} \int \mathrm{d}r r^2 R_n(r) g(r)$$
+        ///
+        /// where $g(r)$ is a radially symmetric density function, $R_n(r)$ the radial
+        /// basis function and $n$ the current radial channel. Note that the integration
+        /// range was deliberately left ambiguous since it depends on the radial basis,
+        /// i.e. for the GTO basis, $r \in \mathbb{R}^+$ is used, while
+        /// $r \in [0, r_\mathrm{cut}]$ for the monomial basis.
+        ///
+        /// This parameter is only required for LODE calculations where the length of
+        /// the provided `center_contribution` must be the same as `max_radial`.
```
Comment (Luthaf) [inline rascaline/src/calculators/radial_basis/mod.rs:45]: Unfortunately, we don't render inner members documentation in the doc, could you move this above in the `TabulatedRadialIntegral` doc? Cf the link in my other comment for a rendered version.

File: rascaline/src/calculators/radial_basis/mod.rs (hunk 5 above)
Comment (PicoCentauri) [reply rascaline/src/calculators/radial_basis/mod.rs:13]: i only checked the rust docs and there it looks fine. Sure I can adjust this.

Comment (Luthaf) [reply rascaline/src/calculators/radial_basis/mod.rs:13]: yeah, this is used in two places. We could (and probably should) improve the calculator reference to render math, but the code extracting the doc is a bit crude.

Comment (PicoCentauri) [reply rascaline/src/calculators/radial_basis/mod.rs:13]: Yeah its fine. Now I know :-)

File: rascaline/src/calculators/radial_basis/mod.rs [hunk 7]
```diff
@@ -27,8 +27,17 @@ pub enum RadialBasis {
...
     /// The easiest way to create a set of spline points is the
     /// `rascaline.generate_splines` Python function.
+    ///
+    /// For LODE calculations also the contribution of the central atom have to be
+    /// provided. The `center_contribution` is defined as `c_n =
+    /// \sqrt{4π} \int dr r^2 R_n(r) g(r)` where `g(r)` is a radially symmetric density
+    /// function, `R_n(r)` the radial basis function and `n` the current radial channel.
+    /// Note that the integration range was deliberately left ambiguous since it depends
+    /// on the radial basis, i.e. for the GTO basis, `r \in R^+` is used, while `r \in
+    /// [0, cutoff]` for the monomial basis.
```
Comment (Luthaf) [inline rascaline/src/calculators/radial_basis/mod.rs:37]: I've seen that a couple of times, but I'm not sure I agree: if you want a smaller integration range, you can just define your basis to be zero outside the cutoff. And then the integration range is always `0, \infty`. I'm fine with leaving this here for now though, we can always revisit this when we update rascaline's hypers

**Review by Luthaf — APPROVED** [review]

**Pull request merged**